File: src/main/java/org/dbshell/commands/connections/DatabaseMdPrimitiveProvider.java
package org.dbshell.commands.connections;

import org.apache.commons.io.FilenameUtils;
import org.bradfordmiller.simplejndiutils.JNDIUtils;
import org.dbshell.environment.ContextAndJndi;
import org.dbshell.environment.EnvironmentVars;
import org.dbshell.reflection.utils.DatabaseMetadataUtil;
import org.springframework.core.MethodParameter;
import org.springframework.shell.CompletionContext;
import org.springframework.shell.CompletionProposal;
import org.springframework.shell.standard.ValueProviderSupport;
import org.springframework.stereotype.Component;

import java.io.File;
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.SQLException;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class DatabaseMdPrimitiveProvider extends ValueProviderSupport {
    @Override
    public List<CompletionProposal> complete(MethodParameter parameter, CompletionContext completionContext, String[] hints) {
        String currentInput = completionContext.currentWordUpToCursor();

        ContextAndJndi ctxJndi = EnvironmentVars.Companion.getCurrentContextAndJndi();
        DatabaseMetaData dbmd = null;
        try {
            Connection connection = JNDIUtils.getJndiConnection(ctxJndi.getJndi(), ctxJndi.getContext());
            dbmd = connection.getMetaData();
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }

        return DatabaseMetadataUtil.getPrimitivesFromDBMetadata(dbmd)
                .keySet()
                .stream()
                .filter(c -> c.contains(currentInput))
                .map(p -> FilenameUtils.removeExtension(new File(p).getName()))
                .map(CompletionProposal::new)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/org/dbshell/providers/DatabaseMdPrimitiveProvider.kt
package org.dbshell.providers

import org.dbshell.connections.ConnectionRepository
import org.dbshell.environment.EnvironmentVars
import org.dbshell.reflection.utils.DatabaseMetadataUtil
import org.springframework.shell.CompletionContext
import org.springframework.shell.CompletionProposal
import org.springframework.shell.standard.ValueProvider
import org.springframework.stereotype.Component

@Component
class DatabaseMdPrimitiveProvider: ValueProvider {
    override fun complete(
        completionContext: CompletionContext
    ): MutableList<CompletionProposal> {

        val currentInput = completionContext.currentWordUpToCursor()

        val currentConnectionName = EnvironmentVars.currentConnectionName
        val dbmd = ConnectionRepository.loadConnection(currentConnectionName).getConnection().metaData

        return DatabaseMetadataUtil.getPrimitivesFromDBMetadata(dbmd)
            .keys
            .filter{c -> c.contains(currentInput!!)}
            .map {cp -> CompletionProposal(cp)}
            .toMutableList()
    }
}